src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +120,273 @@ namespace wasm {
 
 namespace {
 
+// Contains information about a RefCast we want to move to some target.
+struct RefCastInfo {
+  Type type;
+  RefCast::Safety safety;
+
+  // Indicate whether the cast is already at the target. If it is, then
+  // the information is just for determining if a potential cast is less
+  // refined than the existing cast, and will be removed later.
+  bool atTarget;
+
+  RefCastInfo() {}
+
+  RefCastInfo(Type type, RefCast::Safety safety, bool atTarget)
+    : type(type), safety(safety), atTarget(atTarget) {}
+
+  void modify(Type newType, RefCast::Safety newSafety, bool newAtTarget) {
+    type = newType;
+    safety = newSafety;
+    atTarget = newAtTarget;
+  }
+};
+
+// Contains information about a RefAs we want to move to some target.
+struct RefAsInfo {
+  RefAsOp op;
+  bool atTarget;
+
+  RefAsInfo() {}
+
+  RefAsInfo(RefAsOp op, bool atTarget) : op(op), atTarget(atTarget) {}
+};
+
+// Find casts to move earlier to another local.gets. More refined subtypes are
+// chosen over less refined ones.
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+  size_t numLocals;
+
+  // For each index, tracks a list of most refined casts that we can move to the
+  // earliest reachable local.get.
+  std::vector<SmallVector<RefCastInfo, 5>> movedRefCast;
+  std::vector<SmallVector<RefAsInfo, 3>> movedRefAs;
```

Comment:
(why 5 and 3 and not the same value in both?)

Reply:
No particular reason. I picked 3 for `movedRefAs` because there are only 3 members of `RefAsOp`. I just picked 5 for `RefCast`, but 3 is probably also reasonable, since I assume there shouldn't be too many different types. If there is a way to specify this without hard-coding the values, let me know.